modules/ocf_mirrors/files/collect-mirrors-stats

```diff
@@ -0,0 +1,168 @@
+#!/usr/bin/python3
+import argparse
+import os
+import traceback
+from datetime import date
+from datetime import datetime
+from datetime import timedelta
+from pathlib import Path
+
+from ocflib.lab.stats import get_connection
+from ocflib.lab.stats import humanize_bytes
+
+OCFSTATS_PWD = os.environ.get('OCFSTATS_PWD', None)
+MIRRORS_DATA_PATH = Path('/opt/mirrors/ftp')
+APACHE_LOG_PATH = Path('/var/log/apache2')
+APACHE_LOG_FILES = [
+    Path('mirrors.ocf.berkeley.edu_access.log'),
+    Path('mirrors.ocf.berkeley.edu-ssl_access_ssl.log')
+]
+RSYNC_LOG_PATH = Path('/var/log/rsync')
+
+
+def log_to_mysql(projects, dt=None, quiet=False):
+    log_date = dt or date.today()
+
+    with get_connection(user='ocfstats', password=OCFSTATS_PWD) as cursor:
+        for name, data in projects.items():
+            cursor.execute(
+                'INSERT INTO `mirrors` (`date`, `dist`, `up`, `down`) '
+                'VALUES (%s, %s, %s, %s)',
+                (log_date, name, data['up'], data['down'])
+            )
+
+
+def process_apache_log(projects, fn, log_date):
+    """Extract transfer data for a given day for all projects from an Apache log file.
+
+    We iteratively populate projects from all the log files in
+    consideration (4 by default) to avoid reading the entire file
+    repeatedly for multiple projects and to avoid having to merge
+    dictionaries.
+    """
+
+    with fn.open('r') as f:
+        for n, line in enumerate(f, 1):
+            stats = line.split()
+
+            # apache log date format looks like [11/Jul/2017:00:05:16 -0700]
+            # line.split()[3][1:12] = 11/Jul/2017
+            # we need to dump dates that don't match because
+            # logrotate rotates the logs at 6am
+            try:
+                line_date = datetime.strptime(stats[3][1:12], '%d/%b/%Y')
+            except ValueError:
+                print('Error on line {}:\\  {}\\'.format(n, line))
+                traceback.print_exc()
+
+            if line_date.date() != log_date:
+                continue
+
+            # extract dist name from request url
+            # '/debian/pool/main/h/hwdata/...' -> 'debian'
+            dist = stats[6]
+            dist = dist.split('/')[1] if '/' in dist else dist
+
+            # record if we returned http 2xx/3xx
+            if stats[8][0] in ('2', '3') and dist in projects:
```

Comment:
Hmm... This is pretty ugly. What about at the top of this method you give each of the columns that you use in stats a more descriptive variable name.

Reply:
I would recommend writing an regex to match the log format. See e.g. https://stackoverflow.com/a/12544587. You could even do named captures (`(?P<date>...)`) so that you can reference them like `match['date']`. Though maybe that is overly complicated.

Reply:
I'd prefer something like
```python
status_code = int(stats[STATUS_COL])
if status_code >= 200 and status_code < 400 and dist in projects:
    ...
```
This way, if the thing we think is the status code is in fact not the status code, the int conversion will fail instead of just thinking that it was a 400 or something.

Also, why do we increment "other" when we don't get 2xx or 3xx? Seems like we should just ignore the traffic completely.

Reply:
other contains all kinds of 'interesting' data, like directory traversal attempts, people trying to break the server with strange unicode garbage, it's actually a decent amount of data, usually a couple hundred MB a day or something like that.

Reply:
the 'nice' approach leads to a lot of breakage. the `line[8][0]` approach has the benefit of being so simple that when the log line in question isn't valid, it simply dumps it into `other`, whereas trying to parse it as as int in many cases leads to a `ValueError` when there is some unicode garbage in the log line.